guppylang/compiler/core.py

```diff
@@ -68,70 +114,172 @@ def __init__(
         if parsed_func is None:
             return None
         checked_func = ENGINE.get_checked(parsed_func.id)
-        compiled_func = self.build_compiled_def(checked_func.id)
+        compiled_func = self.build_compiled_def(checked_func.id, type_args)
         assert isinstance(compiled_func, CompiledCallableDef)
         return compiled_func
 
     def declare_global_func(
         self,
         const_id: GlobalConstId,
         func_ty: ht.PolyFuncType,
+        mono_args: PartiallyMonomorphizedArgs | None = None,
     ) -> tuple[hf.Function, bool]:
         """
         Creates a function builder for a global function if it doesn't already exist,
         else returns the existing one.
         """
-        if const_id in self.global_funcs:
-            return self.global_funcs[const_id], True
+        if (const_id, mono_args) in self.global_funcs:
+            return self.global_funcs[const_id, mono_args], True
         func = self.module.define_function(
             name=const_id.name,
             input_types=func_ty.body.input,
             output_types=func_ty.body.output,
             type_params=func_ty.params,
         )
-        self.global_funcs[const_id] = func
+        self.global_funcs[const_id, mono_args] = func
         return func, False
 
+    def type_var_to_hugr(self, var: BoundTypeVar) -> ht.Type:
+        """Compiles a bound Guppy type variable into a Hugr type.
+
+        Takes care of performing partial monomorphization as specified in the current
+        context.
+        """
+        if self.current_mono_args is None:
+            # If we're not inside a monomorphized context, just return the Hugr
+            # variable with the same de Bruijn index
+            return ht.Variable(var.idx, var.hugr_bound)
+
+        match self.current_mono_args[var.idx]:
+            # Either we have a decided to monomorphize the corresponding parameter...
+            case TypeArg(ty=ty):
+                return ty.to_hugr(self)
+            # ... or we're still want to be generic in Hugr
+            case None:
+                # But in that case we'll have to down-shift the de-Bruijn index to
+                # account for earlier params that have been monomorphized away
+                down_shift = sum(1 for arg in self.current_mono_args[: var.idx] if arg)
+                return ht.Variable(var.idx - down_shift, var.hugr_bound)
+            case _:
+                raise InternalGuppyError("Invalid monomorphization")
+
+    def const_var_to_hugr(self, var: BoundConstVar) -> ht.TypeArg:
+        """Compiles a bound Guppy constant variable into a Hugr type argument.
+
+        Takes care of performing partial monomorphization as specified in the current
+        context.
+        """
+        if var.ty != nat_type():
+            raise InternalGuppyError(
+                "Tried to convert non-nat const type argument to Hugr. This should "
+                "have been erased."
```

Comment:
```suggestion
                "have been removed by monomorphization."
```
*not* type erasure (that would mean, runtime behaviour does not depend upon the type!)